File: web/src/lib/social-artifacts.test.ts
import { describe, expect, it } from "vitest";

import type { SessionMessage } from "@/lib/api";
import {
  extractSocialArtifacts,
  hasSocialArtifacts,
  isRemoteImage,
} from "./social-artifacts";

function assistant(content: string, timestamp?: number): SessionMessage {
  return { role: "assistant", content, timestamp };
}

function user(content: string): SessionMessage {
  return { role: "user", content };
}

describe("extractSocialArtifacts", () => {
  it("captures the caption from a linkedin-post fenced block", () => {
    const messages = [
      user("Draft me a post about launching Fabric"),
      assistant(
        "Here is your post:\n\n```linkedin-post\nWe just shipped Fabric.\n\nHere is why it matters.\n```\n\nWant tweaks?",
        1_700_000_000,
      ),
    ];

    const artifacts = extractSocialArtifacts(messages);
    expect(artifacts).toHaveLength(1);
    expect(artifacts[0].caption).toBe(
      "We just shipped Fabric.\n\nHere is why it matters.",
    );
    expect(artifacts[0].messageIndex).toBe(1);
    expect(artifacts[0].timestamp).toBe(1_700_000_000);
  });

  it("never reads a user message even if it names the fence", () => {
    const messages = [
      user("Put the result in a ```linkedin-post``` block please"),
    ];
    expect(extractSocialArtifacts(messages)).toHaveLength(0);
    expect(hasSocialArtifacts(messages)).toBe(false);
  });

  it("pulls a markdown image path from the same message", () => {
    const messages = [
      assistant(
        "```linkedin-post\nBig news today.\n```\n\n## Artifacts\n\n![Launch graphic](assets/launch.png)",
      ),
    ];
    const [artifact] = extractSocialArtifacts(messages);
    expect(artifact.caption).toBe("Big news today.");
    expect(artifact.imagePath).toBe("assets/launch.png");
  });

  it("falls back to a bare image path under an Artifacts heading", () => {
    const messages = [
      assistant(
        "```linkedin-post\nA lesson learned.\n```\n\nArtifacts:\n- ./out/post-image.jpg",
      ),
    ];
    const [artifact] = extractSocialArtifacts(messages);
    expect(artifact.imagePath).toBe("./out/post-image.jpg");
  });

  it("returns a null image when none is present", () => {
    const messages = [assistant("```linkedin-post\nText only post.\n```")];
    const [artifact] = extractSocialArtifacts(messages);
    expect(artifact.imagePath).toBeNull();
  });

  it("ignores fenced blocks with an unrelated language", () => {
    const messages = [
      assistant("```python\nprint('not a post')\n```"),
      assistant("Some prose without any fence."),
    ];
    expect(extractSocialArtifacts(messages)).toHaveLength(0);
  });

  it("captures multiple post blocks across the conversation", () => {
    const messages = [
      assistant("```linkedin-post\nDraft one.\n```"),
      user("try another angle"),
      assistant("```linkedin-post\nDraft two.\n```"),
    ];
    const artifacts = extractSocialArtifacts(messages);
    expect(artifacts.map((a) => a.caption)).toEqual([
      "Draft one.",
      "Draft two.",
    ]);
    expect(new Set(artifacts.map((a) => a.id)).size).toBe(2);
  });
});

describe("isRemoteImage", () => {
  it("distinguishes URLs from workspace paths", () => {
    expect(isRemoteImage("https://example.com/a.png")).toBe(true);
    expect(isRemoteImage("assets/launch.png")).toBe(false);
    expect(isRemoteImage("/home/user/out.png")).toBe(false);
  });
});

File: apps/mobile/android/app/src/test/kotlin/io/github/obliviousodin/fabric/mobile/core/SocialArtifactsTest.kt
package io.github.obliviousodin.fabric.mobile.core

import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertNull
import org.junit.Assert.assertTrue
import org.junit.Test

/**
 * Conformance for the Kotlin artifact parser. Mirrors the shared fixture at
 * apps/mobile/contracts/social-extraction-v1.json (validated against the
 * TypeScript implementation in apps/shared/src/social-contract.test.ts).
 */
class SocialArtifactsTest {
    private data class Msg(
        override val role: String,
        override val content: String?,
        override val timestamp: Long? = null,
    ) : SocialSourceMessage

    @Test
    fun capturesCaptionAndMarkdownImage() {
        val artifacts =
            extractSocialArtifacts(
                listOf(
                    Msg("user", "Draft me a launch post."),
                    Msg(
                        "assistant",
                        "Here you go:\n\n```linkedin-post\nWe shipped Fabric.\n\n" +
                            "Here is why it matters.\n```\n\n## Artifacts\n\n" +
                            "![Launch graphic](assets/launch.png)",
                        1_700_000_000L,
                    ),
                ),
            )

        assertEquals(1, artifacts.size)
        assertEquals("We shipped Fabric.\n\nHere is why it matters.", artifacts[0].caption)
        assertEquals("assets/launch.png", artifacts[0].imagePath)
        assertEquals(1, artifacts[0].messageIndex)
        assertEquals(1_700_000_000L, artifacts[0].timestamp)
        assertEquals("1:0", artifacts[0].id)
    }

    @Test
    fun ignoresUserMessageNamingTheFence() {
        val messages = listOf(Msg("user", "Put the result in a ```linkedin-post``` block please."))
        assertTrue(extractSocialArtifacts(messages).isEmpty())
        assertFalse(hasSocialArtifacts(messages))
    }

    @Test
    fun textOnlyPostHasNullImageAndTimestamp() {
        val artifact = extractSocialArtifacts(listOf(Msg("assistant", "```linkedin-post\nText only post.\n```")))[0]
        assertEquals("Text only post.", artifact.caption)
        assertNull(artifact.imagePath)
        assertNull(artifact.timestamp)
    }

    @Test
    fun bareImagePathUnderArtifactsIsNotSwallowedByBullet() {
        val artifact =
            extractSocialArtifacts(
                listOf(Msg("assistant", "```linkedin-post\nA lesson learned.\n```\n\nArtifacts:\n- ./out/post-image.jpg")),
            )[0]
        assertEquals("./out/post-image.jpg", artifact.imagePath)
    }

    @Test
    fun ignoresUnrelatedCodeFence() {
        val artifacts =
            extractSocialArtifacts(
                listOf(
                    Msg("assistant", "```python\nprint('not a post')\n```"),
                    Msg("assistant", "Some prose without any fence."),
                ),
            )
        assertTrue(artifacts.isEmpty())
    }

    @Test
    fun capturesMultipleDraftsInOrder() {
        val artifacts =
            extractSocialArtifacts(
                listOf(
                    Msg("assistant", "```linkedin-post\nDraft one.\n```"),
                    Msg("user", "try another angle"),
                    Msg("assistant", "```linkedin-post\nDraft two.\n```"),
                ),
            )
        assertEquals(listOf("Draft one.", "Draft two."), artifacts.map { it.caption })
        assertEquals(listOf("0:0", "2:0"), artifacts.map { it.id })
    }

    @Test
    fun distinguishesRemoteImages() {
        assertTrue(isRemoteImage("https://example.com/a.png"))
        assertFalse(isRemoteImage("assets/launch.png"))
        assertFalse(isRemoteImage("/home/user/out.png"))
    }
}
